checkov/terraform/checks_infra/solvers/connections_solvers/and_connection_solver.py

```diff
@@ -1,43 +1,36 @@
+from typing import Optional, List, Tuple, Dict, Any
+
 from networkx.classes.digraph import DiGraph
 
 from checkov.common.graph.checks_infra.enums import Operators
+from checkov.common.graph.checks_infra.solvers.base_solver import BaseSolver
 from checkov.terraform.checks_infra.solvers.connections_solvers.complex_connection_solver import ComplexConnectionSolver
 from checkov.terraform.graph_builder.graph_components.attribute_names import CustomAttributes
 
 
 class AndConnectionSolver(ComplexConnectionSolver):
     operator = Operators.AND
 
-    def __init__(self, solvers, operator):
+    def __init__(self, solvers: Optional[List[BaseSolver]], operator: str) -> None:
         super().__init__(solvers, operator)
 
-    def get_operation(self, graph_connector: DiGraph):
-        passed_attributes, failed_attributes = self.run_attribute_solvers(graph_connector)
-        passed_attributes = [p for p in passed_attributes if
-                             p[CustomAttributes.ID] not in [f[CustomAttributes.ID] for f in failed_attributes]]
-        passed, failed = passed_attributes, failed_attributes
-        connection_solvers = self.get_sorted_connection_solvers()
-        passed_connections, failed_connections = [], []
-        for connection_solver in connection_solvers:
-            connection_solver.set_vertices(graph_connector, failed_attributes+failed_connections)
+    def get_operation(self, graph_connector: DiGraph) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
+        passed, failed = self.run_attribute_solvers(graph_connector)
+        passed = [
+            p
+            for p in passed
+            if p[CustomAttributes.ID] not in [f[CustomAttributes.ID] for f in failed]
+        ]
+
+        for connection_solver in self.get_sorted_connection_solvers():
+            connection_solver.set_vertices(graph_connector, failed)
             passed_solver, failed_solver = connection_solver.get_operation(graph_connector)
-            passed_connections.extend(passed_solver)
-            failed_connections.extend(failed_solver)
-            passed_connections = [p for p in passed_connections if
-                      p[CustomAttributes.ID] not in [f[CustomAttributes.ID] for f in failed_connections]]
-
-        passed.extend(passed_connections)
-        failed.extend(failed_connections)
-        passed = [p for p in passed if
-                              p[CustomAttributes.ID] not in [f[CustomAttributes.ID] for f in failed]]
-        passed = [p for p in passed if
-                  p[CustomAttributes.ID] in [pt[CustomAttributes.ID] for pt in passed_attributes] and p[
-                      CustomAttributes.ID] in [pt[CustomAttributes.ID] for pt in passed_connections]]
+            passed.extend(passed_solver)
+            failed.extend(failed_solver)
+
+        passed = [p for p in passed if p[CustomAttributes.ID] not in [f[CustomAttributes.ID] for f in failed]]
```

Comment:
@gruebel Nice!
I see we're using this twice.
1. Can we reuse?
2. Can we compute `[f[CustomAttributes.ID] for f in failed]` once, instead of for every passed vertex?

Reply:
much better, I can just remove the first time it is used 😄

Reply:
@gruebel That's not exactly what I meant... we need it twice. I would however split this:
```python
```

to 
```python
```

Reply:
I adjusted it and had to a bit of extra work